components/notebook-controller/controllers/notebook_controller.go

```diff
@@ -449,6 +473,39 @@ func generateVirtualService(instance *v1beta1.Notebook) (*unstructured.Unstructu
 			"timeout": "300s",
 		},
 	}
+
+	// Create a template map for the headers section
+	// of the Istio VirtualService
+	headers := map[string]interface{}{
+		"request": map[string]interface{}{
+			"set": map[string]interface{}{},
+		},
+	}
+
+	// Blacklist for request headers that are not allowed to be set
+	blacklist := []string{"userid-header", "X-Forwarded-Proto"}
+
+	// If key "notebooks.kubeflow.org/http-headers-request-set" is present on the notebook resource,
+	// split the literal string into keys and values and add them
+	// to the headers variable. Finally, add the headers section to
+	// the variable with the http spec.
+	if _, ok := annotations["notebooks.kubeflow.org/http-headers-request-set"]; ok && len(annotations["notebooks.kubeflow.org/http-headers-request-set"]) > 0 {
```

Comment:
Also NIT, lets use `http-request-headers-set` [ switch the headers/request words ] instead

Reply:
The current naming for these annotations was chosen as this is follows the structure of the resulting VirtualService yaml. 
`notebooks.kubeflow.org/http-rewrite-uri`: sets `http[0].rewrite.uri`
`notebooks.kubeflow.org/http-headers-request-set`: sets `http[0].headers.request.set`

Reply:
ACK. I think it comes more natural to pronounce `http request headers set` rather than `http headers request set`. Could you make this small change?

Reply:
Sure, I could honestly not care less. Pushed the commit.

Reply:
Thank you!

Reply:
The reason for the strange ordering of the words is because it maps to the Istio VirtualService `http[0].headers.request.set`, therefore, for consistency, let's leave it as is.